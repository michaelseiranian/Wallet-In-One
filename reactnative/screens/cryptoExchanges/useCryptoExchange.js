import { useState } from "react";
import * as SecureStore from "expo-secure-store";
import { BACKEND_URL } from "@env"
import { api_url } from '../../authentication';


export default function useCryptoExchange() {
  const [exchanges, setExchanges] = useState([]);
  // Fetch the crypto exchange accounts
  const fetchExchanges = async () => {
    await fetch(`${api_url}/crypto-exchanges`, {
        method: "GET",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Token ${await SecureStore.getItemAsync("token")}`,
        },
      })
        .then((res) => res.json())
        .then((res) => setExchanges(res))
        .catch((err) => console.log(err));
  }
  // Remove exchange backend call
  const removeExchange = async (id) => {
    await fetch(`${api_url}/crypto-exchanges`, {
      method: "DELETE",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Token ${await SecureStore.getItemAsync("token")}`,
      },
      body: JSON.stringify({
        id: id
      }),
    })
      .then(() => setExchanges(exchanges.filter(exchange => exchange.id !== id)))
      .catch((err) => console.log(err));

    }

  return { exchanges, setExchanges, fetchExchanges, removeExchange };
}