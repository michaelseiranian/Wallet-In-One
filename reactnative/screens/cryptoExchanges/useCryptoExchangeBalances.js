import { useState } from "react";
import * as SecureStore from "expo-secure-store";
import { BACKEND_URL } from "@env"
import { api_url } from '../../authentication';

export default function useCryptoExchangeBalances() {
    const [balances, setBalances] = useState([]);
    // Fetch balances
    const fetchBalances = async () => {
      await fetch(`${api_url}/crypto-exchanges/get_exchange_balances/`, {
          method: "GET",
          headers: {
            "Content-Type": "application/json",
            Authorization: `Token ${await SecureStore.getItemAsync("token")}`,
          },
        })
          .then((res) => res.json())
          .then((res) => setBalances(res))
          .catch((err) => console.log(err));
    }
  
    return { balances, setBalances, fetchBalances};
}