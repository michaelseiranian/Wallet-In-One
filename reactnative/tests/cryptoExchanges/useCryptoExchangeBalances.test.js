import { renderHook } from '@testing-library/react-hooks';
import useCryptoExchangeBalances from "../../screens/cryptoExchanges/useCryptoExchangeBalances";
import expect from "expect";
import {describe, test} from "@jest/globals";
import {act} from "@testing-library/react-native";
import useCryptoExchange from "../../screens/cryptoExchanges/useCryptoExchange";


describe('useCryptoExchangeBalances', () => {
    test('should match snapshot', () => {
        const { result } = renderHook(() => useCryptoExchangeBalances());
        expect(result.current).toMatchSnapshot();
    });

    beforeEach(() => {
        jest.clearAllMocks();
    });

    test('fetchBalances works correctly', async () => {
        const mockFetch = jest.fn(() =>
            Promise.resolve({
                json: () => Promise.resolve([
                    {
                        "x": "Binance",
                        "y": 500,
                        "id": 1
                    }
                ]),
            })
        );
        global.fetch = mockFetch;

        const { result } = renderHook(() => useCryptoExchangeBalances());
        expect(result.current.balances).toEqual([]);
        await act( async () => {
            await result.current.fetchBalances();

            expect(mockFetch).toHaveBeenCalledWith(expect.stringContaining(`/crypto-exchanges/get_exchange_balances/`), {
                method: 'GET',
                headers: {
                    'Content-Type': 'application/json',
                    Authorization: `Token undefined`,
                },
            });
            expect(result.current.balances).toEqual([
                {
                    "x": "Binance",
                    "y": 500,
                    "id": 1
                }
            ]);
        });
    });

    test('fetchBalances rejects expectedly', async () => {
        const mockFetch = jest.fn(() => {
            return Promise.reject();
        })

        global.fetch = mockFetch;
        const {result} = renderHook(() => useCryptoExchangeBalances());

        await act(async () => {
            result.current.setBalances([{"x": "Binance", "y": 500, "id": 1}]);

            await result.current.fetchBalances();
        });
        expect(mockFetch).toHaveBeenCalledWith(expect.stringContaining(`/crypto-exchanges/get_exchange_balances/`), {
            method: 'GET',
            headers: {
                'Content-Type': 'application/json',
                Authorization: `Token undefined`,
            }
        });
    });
});