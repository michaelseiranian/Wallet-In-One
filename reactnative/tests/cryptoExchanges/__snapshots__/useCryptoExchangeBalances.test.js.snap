// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCryptoExchangeBalances should match snapshot 1`] = `
{
  "balances": [],
  "fetchBalances": [Function],
  "setBalances": [Function],
}
`;
